### DISCUSSION TITLE
How to add my own pattern?

### DISCUSSION THREAD

**Discussion by polyglot-jones** [danielmiessler/Fabric#167] [Q&A]
I'm just getting started here, directly invoking patterns from the client. I ran a couple of the existing patterns. So far, so good. Next, I tried writing my own, new, pattern. I created a new subfolder within the patterns folder and populated it with a `system.md` file. I then did `fabric -u` followed by `fabric -p my_new_pattern`, but fabric can't find my new pattern. What am I doing wrong? I don't see anything in the readme file on this topic.

**Comment by xssdoctor** [comment]
Did you put your new pattern in .config/fabric/patterns?

**Reply by RoeslerM** [reply]
Ah, that solved also my issue. I thought I have to put that in fabric/pattern.
Does it has to be in both directories ? Then best as symbolic link ?

**Reply by xssdoctor** [reply]
No only in the .config directory

**Reply by cafe99** [reply]
Hi, i will like to write my own pattern, any instruction to create my own pattern and update to fabric.  Thanks.

**Reply by CPAtoCybersecurity** [reply]
Hi @cafe99 , I have a blog post with the steps here and will have a YouTube video walking through it posted by Monday: https://cpatocybersecurity.beehiiv.com/p/custom-patterns.  Hope that helps and if you have any questions I can do my best to try and help.

**Comment by cafe99** [comment]
Hi @CPAtoCybersecurity,  awesome.  Thanks for the help.  It work well..
> Hi @cafe99 , I have a blog post with the steps here and will have a YouTube video walking through it posted by Monday: https://cpatocybersecurity.beehiiv.com/p/custom-patterns. Hope that helps and if you have any questions I can do my best to try and help.

**Reply by CPAtoCybersecurity** [reply]
Great stuff and here's the video.  Feedback welcome. https://youtu.be/7ftiQ8dVvO4

**Comment by nic0711** [comment]
"fabric --update" overrides may custom patterns under /.config/fabric/patterns/ ... what can i do?

**Reply by xssdoctor** [reply]
It shouldn't be doing that as long as your custom patterns have a custom name. I.e. if you edit extract_wisdom --update will overwrite but if you make your_wisdom it should persist. Can you confirm that you have custom names?

**Reply by nic0711** [reply]
They had custom names - now they are gone. ;)

**Reply by xssdoctor** [reply]
Ok let me look into it

**Reply by xssdoctor** [reply]
It working perfectly. but I think I know what your problem is. the patterns are loaded from .config/fabric/patterns not from fabric/patterns. You should completely ignore fabric/patterns in fact. Put your custom patterns in .config/fabric/patterns. Run update after that and please confirm that everything is working

**Reply by nic0711** [reply]
I had done it like this. (Only created in the ./config/... path)
Since I wanted to do it particularly "right", I "obviously" created and saved the patterns exclusively in the folder - without backups, of course <<head-table-repeat>>

I'll try it again with other patterns and then let you know.

**Reply by nic0711** [reply]
... still deleting custom folders

Maybe i need an update of fabric core? ... git clone.. ?
 
 
Update: 
Just saw that: "git pull; pipx upgrade fabric; fabric --update; source ~/.zshrc (or ~/.bashrc)" 

--> zsh: unknown sort specifier

**Reply by xssdoctor** [reply]
Yeah you have to git pull then pipx upgrade fabric

**Reply by nic0711** [reply]
SOLVED:

cd /../fabric 
"git pull; pipx upgrade fabric; fabric --update; source ~/.zshrc (or ~/.bashrc)" did the job

After the update of the fabric core it seems to work now.

**Comment by MaksimZinovev** [comment]
Do I need to run any comments once I added or updated existing new pattern? The latest version will be used automatically?
